---
title: "R/qtl2ggplot Vignette"
author: "Brian S. Yandell"
date: "2/1/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This document examines the MLEs of QTL effects from three basic models (here `allele` signifies founder allele):

* 8 allele means at maximum LOD for additive allele model
* 36 allele pair (diplotype) means
    + at maximum LOD for additive allele model
    + at maximum LOD for allele pair model
    + at maximum LOD for best scan patterns
* 3-level SNPs
    + at maximum LOD for SNPs by allele pattern

This is illustrated with the DOex dataset at <https://github.com/rqtl/qtl2data> with 261 mice. 
For these data,
there is one phenotype with a peak on chr 2 near 100Mbp. However, the peak position varies slightly by model. Key findings:

* LOD profile for 36 allele pair model is unstable (sample size too small)
* 36 allele pair means are unstable, but useful to look at sets of them
* scan patterns are useful but can be counter-intuitive

The traditional QTL effects compare 8 founder additive allele effects, 
while each scan pattern contrast compares the 3 SNP alleles,
reference (`ref` = `dark green`), 
heterozygote (`het` = `dark orange`) and 
alternate (`alt` = `dark blue`).

## Setup

Load example DO data from web

```{r}
library(qtl2)
library(qtl2ggplot)
library(ggplot2)
```

```{r}
dirpath <- "~/Documents/Research/rqtl/data/"
if(!file.exists(file <- file.path(dirpath, "DOex.zip"))) {
  dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                    "qtl2data/master/DOex")
  file <- file.path(dirpath, "DOex.zip")
}
DOex <- read_cross2(file)
```

Subset to chr 2.

```{r}
DOex <- DOex[,"2"]
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Download snp info from web

```{r}
if(!file.exists(file <- file.path(dirpath, "c2_snpinfo.rds"))) {
  dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                    "qtl2data/master/DOex")
  file <- file.path(dirpath, "c2_snpinfo.rds")
  tmpfile <- tempfile()
  download.file(file, tmpfile, quiet=TRUE)
  snpinfo <- readRDS(tmpfile)
  unlink(tmpfile)
} else {
  snpinfo <- readRDS(file)
}
```

Calculate strain distribution patterns

```{r}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
snpinfo <- index_snps(DOex$pmap, snpinfo)
```

Convert to snp probabilities

```{r}
snppr <- genoprob_to_snpprob(apr, snpinfo)
```

## Genome Scans

Additive 8 allele genome scan.

```{r}
scan_apr <- scan1(apr, DOex$pheno)
```

36 allele pair genome scan.

```{r}
scan_pr <- scan1(pr, DOex$pheno)
```

Best patterns from SNP association.

```{r}
out_snppr <- scan1(snppr, DOex$pheno)
```

## QTL Effects

QTL effects for 8 allele model.

```{r}
coefs <- scan1coef(apr, DOex$pheno)
```

```{r}
autoplot(coefs, DOex$pmap)
```

QTL effects for 36 allele pair model. Note that they are quite unstable, and the 36 allele pair max LOD is far from the peak for the additive (haplotype) model. Only showing effects with at least one `E` allele. Plots are truncated at +/-100 for viewability.

```{r}
coefs36 <- scan1coef(pr, DOex$pheno)
```

All 36 allele pair QTL effects.

```{r}
autoplot(coefs36, DOex$pmap, CC = FALSE, ylim=c(-100,100))
```

Only 8 allele pair QTL effects that contain `E`.

```{r}
tmp <- qtl2pattern:::modify_object(coefs36, 
                    coefs36[, stringr::str_detect(dimnames(coefs36)[[2]], "E")])
autoplot(tmp, DOex$pmap, ylim=c(-100,100))
```

```{r}
effects36 <- function(coefs36, scan_apr, map, allele = "E") {
  tmp <- summary(coefs36, scan_apr, map)
  tmp <- tmp[, !stringr::str_detect(names(tmp), paste(LETTERS, collapse="|")) | 
               stringr::str_detect(names(tmp), allele)]
  tidyr::gather(tmp, allele, effect, -pheno, -chr, -pos, -lod)
}
```

```{r}
effects36(coefs36, scan_apr, DOex$pmap)
```

```{r}
effects36(coefs36, scan_pr, DOex$pmap)
```

## Genome Scans

Highlight only top SDP patterns in SNPs.

```{r}
autoplot(out_snppr, snpinfo, patterns="hilit",drop.hilit=1.5,cex=2)
```

```{r}
autoplot(coefs, scan1_output = scan_apr, DOex$pmap)
```
