---
title: "R/qtl2ggplot Vignette"
author: "Brian S. Yandell"
date: "2/1/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This document examines the MLEs of QTL effects in a few ways.

## Setup

Load example DO data from web

```{r}
library(qtl2geno)
library(qtl2scan)
library(qtl2ggplot)
library(qtl2pattern)
```

```{r}
dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                  "qtl2data/master/DOex")
file <- file.path(dirpath, "DOex.zip")
DOex <- read_cross2(file)
```

Subset to chr 2.

```{r}
DOex <- DOex[,"2"]
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Switch map in allele probabilities to Mbp

```{r}
apr$map <- DOex$pmap
pr$map <- DOex$pmap
```

Download snp info from web

```{r}
file <- file.path(dirpath, "c2_snpinfo.rds")
tmpfile <- tempfile()
download.file(file, tmpfile, quiet=TRUE)
snpinfo <- readRDS(tmpfile)
unlink(tmpfile)
```

Calculate strain distribution patterns

```{r}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
```

Convert to snp probabilities

```{r}
snppr <- genoprob_to_snpprob(apr, snpinfo)
```

## Genome Scan

```{r}
out_allele <- scan1(apr, DOex$pheno)
```

```{r}
coefs <- scan1coef(apr, DOex$pheno)
```

```{r}
summary(coefs, out_allele)
```

```{r}
coefs36 <- scan1coef(pr, DOex$pheno)
```

```{r}
summary(coefs36, out_allele)
```

## SNP Association Mapping

Perform SNP association analysis (here, ignoring residual kinship)

```{r}
out_snps <- scan1(snppr, DOex$pheno)
```

```{r}
(patterns <- summary(out_snps))
```

## Scan Patterns

One relatively new and somewhat confusing addition is **scan patterns**, in which the contrast pattern for top SNPs is propagated over the interval to "fill in", leading to LOD and contrast scans. That is, only a few SNPs may show a particular pattern, but key differences might be seen nearby if we could impute SNPs of the same pattern.

```{r}
scan_pat <- qtl2pattern::scan_pattern(pr, DOex$pheno, 
                                      patterns = patterns)
```

```{r}
summary(scan_pat)
```

## Effect estimates

```{r}
alleles <- dplyr::bind_rows(
  haplo = tidyr::gather(
    summary(coefs, out_allele), 
    allele, mean, -pheno, -chr, -pos, -lod),
  diplo = tidyr::gather(
    summary(coefs36, out_allele), 
    allele, mean, -pheno, -chr, -pos, -lod),
  "ABCDGH:EF" = tidyr::gather(
    summary(scan_pat)[1,],
    allele, mean, -pheno, -chr, -pos, -lod),
  "ABCDFGH:E" = tidyr::gather(
    summary(scan_pat)[2,],
    allele, mean, -pheno, -chr, -pos, -lod),
  .id = "source")
```

Pull in `EE` and `alt` for `ABCDGH:EF` by hand

```{r}
tmp <- alleles
tmp$mean[alleles$allele == "EE"] <- alleles$mean[alleles$allele == "E"]
tmp$mean[alleles$allele == "alt" & alleles$source == "ABCDGH:EF"] <- 
  alleles$mean[alleles$allele == "DE"]
tmp$not_winsor <- !((tmp$allele == "EE") | (tmp$allele == "alt" & tmp$source == "ABCDGH:EF"))
ggplot2::ggplot(tmp, 
                ggplot2::aes(x=source, y=mean, label=allele, col=not_winsor)) +
  ggrepel::geom_text_repel()
```

## Genome Scans

Here we show the traditional QTL effects and LOD scan next to the newer scan pattern effects and LOD scans for comparison. The traditional LOD has 7 model degrees of freedom (df), while the scan pattern contrast LODs each have 2 model df. The LOD curves drop off somewhat more steeply for the scan pattern contrasts, and the peak is almost as high (`r round(max(scan_pat$scan)[,3], 1)` vs. `r round(max(out_allele)[,3], 1)`).

The traditional QTL effects compare 8 founder additive allele effects, 
while each scan pattern contrast compares the 3 SNP alleles,
reference (`ref` = `dark green`), 
heterozygote (`het` = `dark orange`) and 
alternate (`alt` = `dark blue`).
The scan pattern contrasts appear at first to be confusing and contradictory, with the `ABCDGH:EF` contrast showing that the `ref` allele is dominant (with `het` at the same level), while the `ABCDFGH:E` suggests `ref` is recessive. However, a closer look shows the following story with regard to the `alt` allele:

* the `ABCDFGH:E` contrast finds 1 or 2 `E=NZO` alleles have lower response
* the `ABCDGH:EF` contrast finds having only `E=NZO` and/or `F=CAST` alleles leads to lower response

These findings are thus complementary, and they agree with the traditional QTL effects plot showing `NZO` = `powder blue` and `CAST` = `green` alleles being much lower than the other 6 alleles.

Highlight only top SDP patterns in SNPs.

```{r}
plot(out_snps, patterns="hilit",drop.hilit=1.5,cex=2)
```

```{r}
plot(coefs, scan1_output = out_allele)
```

```{r}
plot(scan_pat, "coef_and_lod")
```
