---
title: "R/qtl2ggplot Vignette"
author: "Brian S. Yandell"
date: "2/1/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup'

Load example DO data from web

```{r}
library(qtl2geno)
library(qtl2scan)
library(qtl2ggplot)
```

```{r}
file <- paste0("https://raw.githubusercontent.com/rqtl/",
               "qtl2data/master/DOex/DOex.zip")
DOex <- read_cross2(file)
```

Create artificial second phenotype as arcsic sqrt of first one.

```{r}
DOex$pheno <- cbind(DOex$pheno, asin = asin(sqrt(DOex$pheno[,1]/100)))
```

Subset to chr 2.

```{r}
DOex <- DOex[,"2"]
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Switch map in allele probabilities to Mbp

```{r}
apr$map <- DOex$pmap
```

Download snp info from web

```{r}
tmpfile <- tempfile()
file <- paste0("https://raw.githubusercontent.com/rqtl/",
               "qtl2data/master/DOex/c2_snpinfo.rds")
download.file(file, tmpfile, quiet=TRUE)
snpinfo <- readRDS(tmpfile)
unlink(tmpfile)
```

Calculate strain distribution patterns

```{r}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
```

Convert to snp probabilities

```{r}
snppr <- genoprob_to_snpprob(apr, snpinfo)
```

## Genome Scan

```{r}
out_allele <- scan1(apr, DOex$pheno)
```

```{r}
plot(out_allele)
```

```{r}
coefs <- scan1coef(apr, DOex$pheno)
```

```{r}
plot_coefCC(coefs)
```


## SNP Association Mapping

Perform SNP association analysis (here, ignoring residual kinship)

```{r}
out_snps <- scan1(snppr, DOex$pheno)
```

Plot results

```{r}
plot_snpasso(out_snps)
```

Can also just type plot()

```{r}
plot(out_snps)
```

Plot just subset of distinct SNPs

```{r}
plot_snpasso(out_snps, show_all_snps=FALSE)
```

Highlight the top snps (with LOD within 1.5 of max). Show as open circles of size 1.

```{r}
plot(out_snps, drop.hilit=1.5, cex=1, pch=1)
```

Highlight SDP patterns in SNPs within 3 of max; connect with lines.

```{r}
plot(out_snps, patterns="all",drop.hilit=3,cex=2)
```

Highlight only top SDP patterns in SNPs.

```{r}
plot(out_snps, patterns="hilit",drop.hilit=3,cex=2, 
     ylim = c(3.6,6.6))
```

## Multiple phenotypes

### Genome Scans

```{r}
plot(out_allele, 1:2)
```

Following is still in doqtl2, needing to be migrated (or tossed?).

```{r eval=FALSE}
coefs <- doqtl2::listof_scan1coefCC(as.data.frame(DOex$pheno), covar=NULL, probs=apr, K=NULL)
```

```{r eval=FALSE}
library(dplyr)
doqtl2:::plot.listof_scan1coefCC(coefs, out_allele)
```

### SNP Scans

Plot results. Challenges right now:

- show_all_snps=TRUE does not automatically show facets
- patterns works for one lodcolumn only
- facet works, but value not used? need facet to prevent overlap?

```{r}
plot_snpasso(out_snps,1:2, facet="pheno", cex=2, pch=1)
```

```{r}
plot_snpasso(out_snps,1:2,show_all_snps = FALSE, facet="pheno", cex=2, pch=1)
```

```{r}
plot_snpasso(out_snps,1:2,show_all_snps = FALSE, facet=NULL, 
             cex=2, pch=1, col=1:2, col.hilit=2:1,
             drop.hilit=2)
```

Following does not work now. The pheno is now in facet, but as 1,2. But we lost col column.

```{r}
plot_snpasso(out_snps,1:2,show_all_snps = FALSE, 
             drop.hilit = 2,
             cex=2, pch=1)
```

```{r}
plot_snpasso(out_snps, 2, patterns = "all",
             show_all_snps = FALSE, cex=2, pch=1, drop.hilit=2)
```

```{r}
plot_snpasso(out_snps, 1:2, patterns = "all", show_all_snps = FALSE, cex=2, pch=1,
             drop.hilit=3)
```

```{r}
plot_snpasso(out_snps, 1:2, patterns = "hilit", show_all_snps = FALSE, cex=2, pch=1,
             drop.hilit=3, ylim=c(3.6,6.6))
```

The following one is not working.

```{r}
plot_snpasso(out_snps, 1:2, patterns = "hilit", show_all_snps = FALSE, cex=2, pch=1,
             drop.hilit=3, ylim=c(3,7), facet = "group")
```

This does not work for all SNPs.

```{r}
plot_snpasso(out_snps)
```