---
title: "R/qtl2ggplot Vignette"
author: "Brian S. Yandell"
date: "2/1/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

Load example DO data from web

```{r}
library(qtl2geno)
library(qtl2scan)
library(qtl2ggplot)
library(qtl2pattern)
```

```{r}
dirpath <- "~/Documents/Research/rqtl/data/"
if(!file.exists(file <- file.path(dirpath, "DOex.zip"))) {
  dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                    "qtl2data/master/DOex")
  file <- file.path(dirpath, "DOex.zip")
}
DOex <- read_cross2(file)
```

Create artificial second phenotype as arcsic sqrt of first one.

```{r}
DOex$pheno <- cbind(DOex$pheno, asin = asin(sqrt(DOex$pheno[,1]/100)))
DOex$pheno[,"asin"] <- DOex$pheno[,"asin"] *
  sd(DOex$pheno[,"OF_immobile_pct"], na.rm = TRUE) /
  sd(DOex$pheno[,"asin"], na.rm = TRUE)
```

Subset to chr 2.

```{r}
DOex <- DOex[,"2"]
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Download snp info from web

```{r}
if(!file.exists(file <- file.path(dirpath, "c2_snpinfo.rds"))) {
  dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                    "qtl2data/master/DOex")
  file <- file.path(dirpath, "c2_snpinfo.rds")
  tmpfile <- tempfile()
  download.file(file, tmpfile, quiet=TRUE)
  snpinfo <- readRDS(tmpfile)
  unlink(tmpfile)
} else {
  snpinfo <- readRDS(file)
}
```

Calculate strain distribution patterns

```{r}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
```

Create index to non-equivalent set of SNPs.

```{r}
snpinfo <- index_snps(DOex$pmap, snpinfo)
```

Convert to snp probabilities

```{r}
snpinfo <- index_snps(DOex$pmap, snpinfo)
snppr <- genoprob_to_snpprob(apr, snpinfo)
```

## Genome Scan

```{r}
scan_apr <- scan1(apr, DOex$pheno)
```

```{r}
summary(scan_apr, DOex$pmap)
```

```{r}
plot(scan_apr, DOex$pmap)
```

```{r}
coefs <- scan1coef(apr, DOex$pheno)
```

```{r}
summary(coefs, scan_apr, DOex$pmap)
```

```{r}
plot(coefs, DOex$pmap)
```

Plot allele effects over LOD scan.

```{r}
plot(coefs, DOex$pmap, scan1_output = scan_apr)
```

Examine just some of the founder effects, without centering.

```{r}
plot(coefs, DOex$pmap, c(5,8), facet = "geno")
```

```{r}
plot(coefs, DOex$pmap, 4:5, scan1_output = scan_apr)
```


## SNP Association Mapping

Perform SNP association analysis (here, ignoring residual kinship)

```{r}
scan_snppr <- scan1(snppr, DOex$pheno)
```

```{r}
summary(scan_snppr, DOex$pmap, snpinfo)
```

Plot results

```{r}
plot(scan_snppr, snpinfo, lodcolumn=1:2, facet = "pheno")
```

Plot just subset of distinct SNPs

```{r}
plot(scan_snppr, snpinfo, show_all_snps=FALSE)
```

Highlight the top snps (with LOD within 1.5 of max). Show as open circles of size 1.

```{r}
plot(scan_snppr, snpinfo, drop.hilit=1.5, cex=1, pch=1)
```

Highlight SDP patterns in SNPs within 3 of max; connect with lines.

```{r}
plot(scan_snppr, snpinfo, patterns="all",drop.hilit=3,cex=2)
```

Highlight only top SDP patterns in SNPs.

```{r}
plot(scan_snppr, snpinfo, patterns="hilit",drop.hilit=3,cex=2,
     ylim = c(3.6,6.6))
```

## Scan Patterns

One relatively new and somewhat confusing addition is **scan patterns**, in which the contrast pattern for top SNPs is propagated over the interval to "fill in", leading to LOD and contrast scans. That is, only a few SNPs may show a particular pattern, but key differences might be seen nearby if we could impute SNPs of the same pattern.

```{r}
top_snps_tbl <- qtl2pattern::top_snps_all(scan_snppr, snpinfo)
```

```{r}
(patterns <- dplyr::arrange(
  summary(
    dplyr::rename(top_snps_tbl, snp=snp_id)),
  dplyr::desc(max_lod)))
```

```{r}
(pheno_name <- dimnames(DOex$pheno)[[2]][1])
```

```{r}
scan_pat <- qtl2pattern::scan_pattern(pr, DOex$pheno,
                                      map = DOex$pmap,
                                      patterns = patterns)
```

```{r}
summary(scan_pat, DOex$pmap)
```

Modify plot_scan_pattern to facet on multiple phenotypes.

```{r}
plot(scan_pat, DOex$pmap) +
  ggplot2::geom_vline(xintercept = c(96.5,98.5))
```

```{r}
plot(scan_pat, DOex$pmap, "coef", ylim=c(-100,100)) +
  ggplot2::geom_vline(xintercept = c(96.5,98.5))
```

The following does not work. Points to multiple use of `pattern` and `patterns` in routines.

```{r eval=FALSE}
plot(scan_pat, DOex$pmap, "coef_and_lod", ylim=c(-100,100)) +
  ggplot2::geom_vline(xintercept = c(96.5,98.5))
```

Here we show the traditional QTL effects and LOD scan next to the newer scan pattern effects and LOD scans for comparison. The traditional LOD has 7 model degrees of freedom (df), while the scan pattern contrast LODs each have 2 model df. The LOD curves drop off somewhat more steeply for the scan pattern contrasts, and the peak is almost as high (`r round(max(scan_pat$scan, DOex$pmap)[,3], 1)` vs. `r round(max(scan_apr, DOex$pmap)[,3], 1)`).

The traditional QTL effects compare 8 founder additive allele effects,
while each scan pattern contrast compares the 3 SNP alleles,
reference (`ref` = `dark green`),
heterozygote (`het` = `dark orange`) and
alternate (`alt` = `dark blue`).
The scan pattern contrasts appear at first to be confusing and contradictory, with the `ABCDGH:EF` contrast showing that the `ref` allele is dominant (with `het` at the same level), while the `ABCDFGH:E` suggests `ref` is recessive. However, a closer look shows the following story with regard to the `alt` allele:

* the `ABCDFGH:E` contrast finds 1 or 2 `E=NZO` alleles have lower response
* the `ABCDGH:EF` contrast finds having only `E=NZO` and/or `F=CAST` alleles leads to lower response

These findings are thus complementary, and they agree with the traditional QTL effects plot showing `NZO` = `powder blue` and `CAST` = `green` alleles being much lower than the other 6 alleles.

```{r}
plot(coefs, DOex$pmap, scan1_output = scan_apr)
```

```{r}
plot(scan_pat, DOex$pmap, "coef_and_lod", ylim = c(-100,100))
```

## Multiple phenotypes

Plot routines (except scan patterns for now) can accommodate multiple phenotypes. At present, it is best to stick to under 10. In the preambl of this document, a second phenotype, `asin`, was artifically created for illustration purposes.

### Genome Scans

```{r}
plot(scan_apr, DOex$pmap, 1:2)
```

```{r}
plot(scan_apr, DOex$pmap, 1:2, facet="pheno")
```

```{r}
coefs2 <- qtl2pattern::listof_scan1coef(apr, DOex$pheno)
```

Experiment with coefs

```{r}
plot(coefs2, DOex$pmap)
```

```{r}
plot(coefs2, DOex$pmap, columns=4:5)
```

```{r}
plot(coefs2, DOex$pmap, facet="geno")
```

```{r}
plot(coefs2, DOex$pmap, 4:5, facet="geno")
```

### SNP Scans

Plot results. Challenges right now:

```{r}
plot(scan_snppr, snpinfo, 1:2, facet="pheno", cex=1, pch=1)
```

```{r}
plot(scan_snppr, snpinfo, 1:2, show_all_snps = FALSE, facet="pheno", cex=2, pch=1)
```

```{r}
plot(scan_snppr, snpinfo, 1:2,show_all_snps = FALSE,
             drop.hilit = 2, col=1:2, col.hilit=3:4,
             cex=2, pch=1)
```

Play with the colors.

```{r}
plot(scan_snppr, snpinfo, 1:2,show_all_snps = FALSE, facet_var = "pheno",
             drop.hilit = 2, col=1:2, col.hilit=2:1,
             cex=2, pch=1)
```

```{r}
plot(scan_snppr, snpinfo, 2, patterns = "all",
             cex=2, pch=1, drop.hilit=2)
```

```{r}
plot(scan_snppr, snpinfo, 1:2, patterns = "all", cex=2, pch=1,
             facet = "pheno", drop.hilit=3)
```

```{r}
plot(scan_snppr, snpinfo, 1:2, patterns = "hilit", cex=2, pch=1,
             drop.hilit=3, ylim=c(3.6,6.6), facet = "pheno")
```

```{r}
plot(scan_snppr, snpinfo, 1:2, patterns = "hilit", show_all_snps = TRUE, cex=2, pch=1,
             drop.hilit=3, ylim=c(3,7), facet = "pattern")
```
