---
title: "R/qtl2ggplot Vignette"
author: "Brian S. Yandell"
date: "2/1/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup'

Load example DO data from web

```{r}
library(qtl2geno)
library(qtl2scan)
library(qtl2ggplot)
```

```{r}
dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                  "qtl2data/master/DOex")
file <- file.path(dirpath, "DOex.zip")
DOex <- read_cross2(file)
```

Create artificial second phenotype as arcsic sqrt of first one.

```{r}
DOex$pheno <- cbind(DOex$pheno, asin = asin(sqrt(DOex$pheno[,1]/100)))
DOex$pheno[,"asin"] <- DOex$pheno[,"asin"] * 
  sd(DOex$pheno[,"OF_immobile_pct"], na.rm = TRUE) /
  sd(DOex$pheno[,"asin"], na.rm = TRUE)
```

Subset to chr 2.

```{r}
DOex <- DOex[,"2"]
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Switch map in allele probabilities to Mbp

```{r}
apr$map <- DOex$pmap
```

Download snp info from web

```{r}
file <- file.path(dirpath, "c2_snpinfo.rds")
tmpfile <- tempfile()
download.file(file, tmpfile, quiet=TRUE)
snpinfo <- readRDS(tmpfile)
unlink(tmpfile)
```

Calculate strain distribution patterns

```{r}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
```

Convert to snp probabilities

```{r}
snppr <- genoprob_to_snpprob(apr, snpinfo)
```

## Genome Scan

```{r}
out_allele <- scan1(apr, DOex$pheno)
```

```{r}
qtl2pattern::summary_scan1(out_allele)
```

```{r}
plot(out_allele)
```

```{r}
coefs <- scan1coef(apr, DOex$pheno)
```

```{r}
summary(coefs, out_allele)
```

```{r}
plot_coefCC(coefs)
```

```{r}
plot(coefs)
```

Plot allele effects over LOD scan.

```{r}
plot(coefs, scan1_output = out_allele)
```

Examine just some of the founder effects, without centering.

```{r}
plot(coefs, c(5,8), facet = "geno", center=F)
```

```{r}
plot(coefs, 4:5, scan1_output = out_allele)
```


## SNP Association Mapping

Perform SNP association analysis (here, ignoring residual kinship)

```{r}
out_snps <- scan1(snppr, DOex$pheno)
```

```{r}
qtl2pattern::summary_scan1(out_snps)
```

Plot results

```{r}
plot(out_snps)
```

Plot just subset of distinct SNPs

```{r}
plot(out_snps, show_all_snps=FALSE)
```

Highlight the top snps (with LOD within 1.5 of max). Show as open circles of size 1.

```{r}
plot(out_snps, drop.hilit=1.5, cex=1, pch=1)
```

Highlight SDP patterns in SNPs within 3 of max; connect with lines.

```{r}
plot(out_snps, patterns="all",drop.hilit=3,cex=2)
```

Highlight only top SDP patterns in SNPs.

```{r}
plot(out_snps, patterns="hilit",drop.hilit=3,cex=2, 
     ylim = c(3.6,6.6))
```

## Multiple phenotypes

### Genome Scans

```{r}
plot(out_allele, 1:2)
```

```{r}
plot(out_allele, 1:2, facet="pheno")
```

Following is still in doqtl2, needing to be migrated (or tossed?).

```{r}
coefs2 <- qtl2pattern::listof_scan1coef(apr, DOex$pheno)
```

Experiment with coefs

```{r}
plot(coefs2)
```

```{r}
plot(coefs2, columns=4:5)
```

```{r}
plot(coefs2, facet="geno")
```

```{r}
plot(coefs2, 4:5, facet="geno")
```

### SNP Scans

Plot results. Challenges right now:

```{r}
plot(out_snps,1:2, facet="pheno", cex=1, pch=1)
```

```{r}
plot(out_snps,1:2,show_all_snps = FALSE, facet="pheno", cex=2, pch=1)
```

```{r}
plot(out_snps,1:2,show_all_snps = FALSE, 
             drop.hilit = 2, col=1:2, col.hilit=3:4,
             cex=2, pch=1)
```

Play with the colors.

```{r}
plot(out_snps,1:2,show_all_snps = FALSE, facet_var = "pheno",
             drop.hilit = 2, col=1:2, col.hilit=2:1,
             cex=2, pch=1)
```

```{r}
plot(out_snps, 2, patterns = "all",
             cex=2, pch=1, drop.hilit=2)
```

```{r}
plot(out_snps, 1:2, patterns = "all", cex=2, pch=1,
             facet = "pheno", drop.hilit=3)
```

```{r}
plot(out_snps, 1:2, patterns = "hilit", cex=2, pch=1,
             drop.hilit=3, ylim=c(3.6,6.6), facet = "pheno")
```

```{r}
plot(out_snps, 1:2, patterns = "hilit", show_all_snps = TRUE, cex=2, pch=1,
             drop.hilit=3, ylim=c(3,7), facet = "pattern")
```
