---
title: "R/qtl2ggplot Vignette"
author: "Brian S. Yandell"
date: "3 December 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

Load example DO data from web

```{r}
library(qtl2)
library(qtl2ggplot)
library(ggplot2)
```

```{r}
if(!file.exists("DOex.zip")) {
  dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                    "qtl2data/master/DOex")
}
DOex <- read_cross2("DOex.zip")
```

Create artificial second phenotype as arcsic sqrt of first one.

```{r}
DOex$pheno <- cbind(DOex$pheno, asin = asin(sqrt(DOex$pheno[,1]/100)))
DOex$pheno[,"asin"] <- DOex$pheno[,"asin"] *
  sd(DOex$pheno[,"OF_immobile_pct"], na.rm = TRUE) /
  sd(DOex$pheno[,"asin"], na.rm = TRUE)
```

Subset to chr 2.

```{r}
DOex <- DOex[,"2"]
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Download snp info from web

```{r}
if(!file.exists(filename <- "c2_snpinfo.rds")) {
  filename <- file.path("https://raw.githubusercontent.com/rqtl",
                        "qtl2data/master/DOex", 
                        "c2_snpinfo.rds")
  tmpfile <- tempfile()
  download.file(filename, tmpfile, quiet=TRUE)
  snpinfo <- readRDS(tmpfile)
  unlink(tmpfile)
} else {
  snpinfo <- readRDS(filename)
}
```

Calculate strain distribution patterns

```{r}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
```

Create index to non-equivalent set of SNPs.

```{r}
snpinfo <- index_snps(DOex$pmap, snpinfo)
```

Convert to snp probabilities

```{r}
snpinfo <- index_snps(DOex$pmap, snpinfo)
snppr <- genoprob_to_snpprob(apr, snpinfo)
```

## Genome Scan

```{r}
scan_apr <- scan1(apr, DOex$pheno)
```

```{r}
find_peaks(scan_apr, DOex$pmap)
```

```{r}
autoplot(scan_apr, DOex$pmap)
```

```{r}
coefs <- scan1coef(apr, DOex$pheno)
```

```{r}
autoplot(coefs, DOex$pmap)
```

Plot allele effects over LOD scan.

```{r}
autoplot(coefs, DOex$pmap, scan1_output = scan_apr)
```

Examine just some of the founder effects, without centering.

```{r}
autoplot(coefs, DOex$pmap, c(5,8), facet = "geno")
```

```{r}
autoplot(coefs, DOex$pmap, 4:5, scan1_output = scan_apr)
```


## SNP Association Mapping

Perform SNP association analysis (here, ignoring residual kinship)

```{r}
scan_snppr <- scan1(snppr, DOex$pheno)
```

Plot results

```{r}
autoplot(scan_snppr, snpinfo, lodcolumn=1:2, facet = "pheno")
```

Plot just subset of distinct SNPs

```{r}
autoplot(scan_snppr, snpinfo, show_all_snps=FALSE)
```

Highlight the top snps (with LOD within 1.5 of max). Show as open circles of size 1.

```{r}
autoplot(scan_snppr, snpinfo, drop.hilit=1.5, cex=1, pch=1)
```

Highlight SDP patterns in SNPs within 3 of max; connect with lines.

```{r}
autoplot(scan_snppr, snpinfo, patterns="all",drop.hilit=3,cex=2)
```

Highlight only top SDP patterns in SNPs.

```{r}
autoplot(scan_snppr, snpinfo, patterns="hilit",drop.hilit=3,cex=2,
     ylim = c(3.6,6.6))
```

## Multiple phenotypes

Plot routines (except scan patterns for now) can accommodate multiple phenotypes. At present, it is best to stick to under 10. In the preambl of this document, a second phenotype, `asin`, was artifically created for illustration purposes.

### Genome Scans

```{r}
autoplot(scan_apr, DOex$pmap, 1:2)
```

```{r}
autoplot(scan_apr, DOex$pmap, 1:2, facet="pheno")
```

### SNP Scans

Plot results. Challenges right now:

```{r}
autoplot(scan_snppr, snpinfo, 1:2, facet="pheno", cex=1, pch=1)
```

```{r}
autoplot(scan_snppr, snpinfo, 1:2, show_all_snps = FALSE, facet="pheno", cex=2, pch=1)
```

```{r}
autoplot(scan_snppr, snpinfo, 1:2,show_all_snps = FALSE,
             drop.hilit = 2, col=1:2, col.hilit=3:4,
             cex=2, pch=1)
```

Play with the colors.

```{r}
autoplot(scan_snppr, snpinfo, 1:2,show_all_snps = FALSE, facet_var = "pheno",
             drop.hilit = 2, col=1:2, col.hilit=2:1,
             cex=2, pch=1)
```

```{r}
autoplot(scan_snppr, snpinfo, 2, patterns = "all",
             cex=2, pch=1, drop.hilit=2)
```

```{r}
autoplot(scan_snppr, snpinfo, 1:2, patterns = "all", cex=2, pch=1,
             facet = "pheno", drop.hilit=3)
```

```{r}
autoplot(scan_snppr, snpinfo, 1:2, patterns = "hilit", cex=2, pch=1,
             drop.hilit=3, ylim=c(3.6,6.6), facet = "pheno")
```

```{r}
autoplot(scan_snppr, snpinfo, 1:2, patterns = "hilit", show_all_snps = TRUE, cex=2, pch=1,
             drop.hilit=3, ylim=c(3,7), facet = "pattern")
```
